File: 백준/Gold/11657. 타임머신/타임머신.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    static final long INF = Long.MAX_VALUE;

    static class Edge {
        int from;
        int to;
        int time;

        Edge(int from, int to, int time) {
            this.from = from;
            this.to = to;
            this.time = time;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        List<Edge> edges = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int time = Integer.parseInt(st.nextToken());
            edges.add(new Edge(from, to, time));
        }

        long[] distance = new long[n + 1];
        Arrays.fill(distance, INF);
        distance[1] = 0;

        for (int i = 1; i < n; i++) {
            for (Edge edge : edges) {
                if (distance[edge.from] != INF && distance[edge.to] > distance[edge.from] + edge.time) {
                    distance[edge.to] = distance[edge.from] + edge.time;
                }
            }
        }

        boolean hasNegativeCycle = false;
        for (Edge edge : edges) {
            if (distance[edge.from] != INF && distance[edge.to] > distance[edge.from] + edge.time) {
                hasNegativeCycle = true;
                break;
            }
        }

        StringBuilder sb = new StringBuilder();
        if (hasNegativeCycle) {
            sb.append(-1);
        } else {
            for (int i = 2; i <= n; i++) {
                if (distance[i] == INF) {
                    sb.append(-1).append('\n');
                } else {
                    sb.append(distance[i]).append('\n');
                }
            }
        }

        System.out.print(sb);
    }
}

File: 백준/Gold/11657. 타임머신/타임머신.kt
import java.io.BufferedReader
import java.io.InputStreamReader

private val inf = Long.MAX_VALUE

data class Edge(val from: Int, val to: Int, val time: Int)

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m) = br.readLine().split(" ").map { it.toInt() }
    val edges = List(m) {
        val (from, to, time) = br.readLine().split(" ").map { it.toInt() }
        Edge(from, to, time)
    }

    val distance = LongArray(n + 1) { inf }
    distance[1] = 0

    repeat(n - 1) {
        for (edge in edges) {
            if (distance[edge.from] != inf && distance[edge.to] > distance[edge.from] + edge.time) {
                distance[edge.to] = distance[edge.from] + edge.time
            }
        }
    }

    val hasNegativeCycle = edges.any { edge ->
        distance[edge.from] != inf && distance[edge.to] > distance[edge.from] + edge.time
    }

    val sb = StringBuilder()
    if (hasNegativeCycle) {
        sb.append(-1)
    } else {
        for (i in 2..n) {
            if (distance[i] == inf) {
                sb.append(-1).append('\n')
            } else {
                sb.append(distance[i]).append('\n')
            }
        }
    }

    print(sb)
}
